python/ray/data/datasource/image_folder_datasource.py

```diff
@@ -0,0 +1,135 @@
+
+    This datasource works with any dataset where images are arranged in this way:
+
+    .. code-block::
+
+        root/dog/xxx.png
+        root/dog/xxy.png
+        root/dog/[...]/xxz.png
+
+        root/cat/123.png
+        root/cat/nsdf3.png
+        root/cat/[...]/asd932_.png
+
+    Datasets read with ``ImageFolderDatasource`` contain two columns: ``'image'`` and
+    ``'label'``. The ``'image'`` column contains ``ndarray`` objects of shape 
+    :math:`(H, W, C)`, and the ``label`` column contains strings corresponding to
+    labels.
+
+    Examples:
+        >>> import ray
+        >>> from ray.data.datasource import ImageFolderDatasource
+        >>>
+        >>> ds = ray.data.read_datasource(  # doctest: +SKIP
+        ...     ImageFolderDatasource(),
+        ...     paths=["/data/imagenet/train"]
+        ... )
+        >>> sample = ds.take(1)[0]  # doctest: +SKIP
+        >>> sample["image"].shape  # doctest: +SKIP
+        (469, 387, 3)
+        >>> sample["label"]  # doctest: +SKIP
+        'n01443537'
+
+    Raises:
+        ValueError: if more than one path is provided. You should only provide the path
+            to the dataset root.
+    """
+
+    def prepare_read(
+        self,
+        parallelism: int,
+        paths: Union[str, List[str]],
+        filesystem: Optional["pyarrow.fs.FileSystem"] = None,
+        schema: Optional[Union[type, "pyarrow.lib.Schema"]] = None,
+        open_stream_args: Optional[Dict[str, Any]] = None,
+        meta_provider: BaseFileMetadataProvider = DefaultFileMetadataProvider(),
+        partition_filter: PathPartitionFilter = None,
+        # TODO(ekl) deprecate this once read fusion is available.
+        _block_udf: Optional[Callable[[Block], Block]] = None,
+        **reader_args,
+    ) -> List[ReadTask]:
+        if len(paths) != 1:
+            raise ValueError(
+                "`ImageFolderDatasource` expects 1 path representing the dataset "
+                f"root, but it got {len(paths)} paths instead. To fix this error, "
+                "pass in a single-element list containing the dataset root (for "
+                'example, `paths=["s3://imagenet/train"]`)'
+            )
+
+        try:
+            import imageio  # noqa: F401
+        except ImportError:
+            raise ValueError(
+                "`ImageFolderDatasource` depends on 'imageio', but 'imageio' couldn't "
+                "be imported. You can install 'imageio' by running "
+                "`pip install imageio`."
+            )
+
+        # We call `_resolve_paths_and_filesystem` so that the dataset root is formatted
+        # in the same way as the paths passed to `_get_class_from_path`.
+        paths, filesystem = _resolve_paths_and_filesystem(paths, filesystem)
+        self.root = paths[0]
+
+        paths, _ = meta_provider.expand_paths(paths, filesystem)
+        paths = [path for path in paths if _is_image_file(path)]
+
+        return super().prepare_read(
```

Comment:
> At a high level, I think this PR can be considerably simplified if you wait for @clarkzinzow 's PR adding file-extension filters to merge: https://github.com/ray-project/ray/pull/24822

I agree. Since this PR isn't particularly time sensitive, I'll wait for #24822 to get merged.